File: extensions/CData ADO.NET Provider for MongoDB 2015/demos - winform/querybrowser/cs/Connection.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data.CData.MongoDB;
using System.Data;
using System.Windows.Forms;
using System.Collections;
using System.ComponentModel;

namespace CData.MongoDB.Demo {

  class QueryData {
    private string _sql;
    private ConnectionNode _connection = null;
    private MongoDBDataAdapter _adapter = null;
    private MongoDBCommandBuilder _updateCommandBuilder = null;
    private MongoDBConnection _conn = null;
    public QueryData(ConnectionNode connection, string sql) {
    this._connection = connection;
    this._sql = sql;
  }

    public DataTable GetData() {
      MongoDBConnection conn = _connection.Connection;
      MongoDBCommand cmd = conn.CreateCommand();
      cmd.CommandText = this._sql;
      cmd.CommandType = CommandType.Text;
      this._adapter = new MongoDBDataAdapter(cmd);
      this._updateCommandBuilder = new MongoDBCommandBuilder(this._adapter);
      DataTable result = new DataTable();
      if (_sql.ToUpper().TrimStart().StartsWith("SELECT")) {
        this._adapter.Fill(result);
      } else {
        int rowsAffected = cmd.ExecuteNonQuery();
        result.Columns.Add("Rows Affected");
        DataRow row = result.NewRow();
        row[0] = rowsAffected;
        result.Rows.Add(row);
      }
      return result;
    }

    public DataTable ExecSP(Hashtable input, Hashtable output, Hashtable inputOutput, Hashtable returnValue) {
      MongoDBConnection conn = _connection.Connection;
      MongoDBCommand cmd = conn.CreateCommand();
      cmd.CommandText = this._sql;
      cmd.CommandType = CommandType.StoredProcedure;
      SetParameters(cmd, input, output, inputOutput, returnValue);
      this._adapter = new MongoDBDataAdapter(cmd);
      DataTable result = new DataTable(this._sql);
      this._adapter.Fill(result);
      GetOutput(cmd, output, inputOutput, returnValue);
      return result;
    }

    public void Update(DataTable data) {
      this._adapter.InsertCommand = this._updateCommandBuilder.GetInsertCommand();
      this._adapter.UpdateCommand = this._updateCommandBuilder.GetUpdateCommand();
      this._adapter.DeleteCommand = this._updateCommandBuilder.GetDeleteCommand();
      this._adapter.Update(data);
    }

    private void SetParameters(MongoDBCommand cmd, Hashtable input, Hashtable output, Hashtable inputOutput, Hashtable returnValue) {
      foreach (object key in input.Keys) {
        MongoDBParameter param = cmd.Parameters.Add(key.ToString());
        param.Value = input[key];
        param.Direction = ParameterDirection.Input;
      }
      foreach (object key in output.Keys) {
        MongoDBParameter param = cmd.Parameters.Add(key.ToString());
        param.Value = output[key];
        param.Direction = ParameterDirection.Output;
      }
      foreach (object key in inputOutput.Keys) {
        MongoDBParameter param = cmd.Parameters.Add(key.ToString());
        param.Value = inputOutput[key];
        param.Direction = ParameterDirection.InputOutput;
      }
      foreach (object key in returnValue.Keys) {
        MongoDBParameter param = cmd.Parameters.Add(key.ToString());
        param.Value = returnValue[key];
        param.Direction = ParameterDirection.ReturnValue;
      }
    }

    private void GetOutput(MongoDBCommand cmd, Hashtable output, Hashtable inputOutput, Hashtable returnValue) {
      foreach (MongoDBParameter param in cmd.Parameters) {
        if (param.Direction == ParameterDirection.Output) {
          output[param.ParameterName] = param.Value;
        } else if (param.Direction == ParameterDirection.InputOutput) {
          inputOutput[param.ParameterName] = param.Value;
        } else if (param.Direction == ParameterDirection.ReturnValue) {
          returnValue[param.ParameterName] = param.Value;
        }
      }
    }
  }
}

File: extensions/CData ADO.NET Provider for MongoDB 2015/demos - winform/querybrowser/vb/Connection.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Data.CData.MongoDB
Imports System.Data
Imports System.Windows.Forms
Imports System.Collections
Imports System.ComponentModel

Namespace CData.MongoDB.Demo
 
  Class QueryData
    Private _connection As ConnectionNode
    Private _sql As String
    Private _adapter As MongoDBDataAdapter = Nothing
    Private _updateCommandBuilder As MongoDBCommandBuilder = Nothing
    Private _conn As MongoDBConnection = Nothing
    Public Sub New(connection As ConnectionNode, sql As String)
      Me._connection = connection
      Me._sql = sql
    End Sub

    Public Function GetData() As DataTable
      Dim conn As MongoDBConnection = _connection.Connection
      Dim cmd As MongoDBCommand = conn.CreateCommand()
      cmd.CommandText = Me._sql
      cmd.CommandType = CommandType.Text
      Me._adapter = New MongoDBDataAdapter(cmd)
      Me._updateCommandBuilder = New MongoDBCommandBuilder(Me._adapter)
      Dim result As New DataTable()
      If Me._sql.ToUpper().TrimStart().StartsWith("SELECT") Then
        Me._adapter.Fill(result)
      Else
        Dim rowsAffected As Integer = cmd.ExecuteNonQuery() 
        result.Columns.Add("Rows Affected")
        Dim row As DataRow =  result.NewRow() 
        row(0) = rowsAffected
        result.Rows.Add(row)
      End If
      Return result
    End Function

    Public Function ExecSP(input As Hashtable, output As Hashtable, inputOutput As Hashtable, returnValue As Hashtable) As DataTable
      Dim conn As MongoDBConnection = _connection.Connection
      Dim cmd As MongoDBCommand = conn.CreateCommand()
      cmd.CommandText = Me._sql
      cmd.CommandType = CommandType.StoredProcedure
      SetParameters(cmd, input, output, inputOutput, returnValue)
      Me._adapter = New MongoDBDataAdapter(cmd)
      Dim result As New DataTable(Me._sql)
      Me._adapter.Fill(result)
      GetOutput(cmd, output, inputOutput, returnValue)
      Return result
    End Function

    Public Sub Update(data As DataTable)
      Me._adapter.InsertCommand = Me._updateCommandBuilder.GetInsertCommand()
      Me._adapter.UpdateCommand = Me._updateCommandBuilder.GetUpdateCommand()
      Me._adapter.DeleteCommand = Me._updateCommandBuilder.GetDeleteCommand()
      Me._adapter.Update(data)
    End Sub

    Private Sub SetParameters(cmd As MongoDBCommand, input As Hashtable, output As Hashtable, inputOutput As Hashtable, returnValue As Hashtable)
      For Each key As Object In input.Keys
        Dim param As MongoDBParameter = cmd.Parameters.Add(key.ToString())
        param.Value = input(key)
        param.Direction = ParameterDirection.Input
      Next
      For Each key As Object In output.Keys
        Dim param As MongoDBParameter = cmd.Parameters.Add(key.ToString())
        param.Value = output(key)
        param.Direction = ParameterDirection.Output
      Next
      For Each key As Object In inputOutput.Keys
        Dim param As MongoDBParameter = cmd.Parameters.Add(key.ToString())
        param.Value = inputOutput(key)
        param.Direction = ParameterDirection.InputOutput
      Next
      For Each key As Object In returnValue.Keys
        Dim param As MongoDBParameter = cmd.Parameters.Add(key.ToString())
        param.Value = returnValue(key)
        param.Direction = ParameterDirection.ReturnValue
      Next
    End Sub

    Private Sub GetOutput(cmd As MongoDBCommand, output As Hashtable, inputOutput As Hashtable, returnValue As Hashtable)
      For Each param As MongoDBParameter In cmd.Parameters
        If param.Direction = ParameterDirection.Output Then
          output(param.ParameterName) = param.Value
        ElseIf param.Direction = ParameterDirection.InputOutput Then
          inputOutput(param.ParameterName) = param.Value
        ElseIf param.Direction = ParameterDirection.ReturnValue Then
          returnValue(param.ParameterName) = param.Value
        End If
      Next
    End Sub
  End Class
End Namespace